### PR TITLE
SolidJS widget to format date[time]s

### PR DESCRIPTION
**Pull request by acarapetis**
Here's my attempt at #42: a SolidJS component FormatDate that can render as an absolute date, absolute datetime or a relative time ago. I've replaced all usage of TimeAgo with it, and also updated MatchPreview to use it.

I've got one annoyign bug hanging around - when using a non-relative format, the rendered HTML is not updating currently when hydrating on the client - the UTC times just hang around forever. The `console.log` debugging call I've got in there is outputting what I expect on the client, so the code is definitely running; it's just not rendering correctly for some reason.

This is my first time touching both Astro and SolidJS so I figured I'd ask for help rather than spend any more time banging my head against it - there's clearly something I don't understand either in how the SSR hydration works, or just how SolidJS works.  Also open to any other feedback :)

### CODE REVIEW

File: src/components/widgets/FormatDate.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+export interface FormatDateProps {
+  date: Date | string
+  format: DateFormat
+}
+export default function FormatDate({ date, format }: FormatDateProps) {
+  const d = parseUTC(date)
+  const [output, setOutput] = createSignal<string>(formatDate(d, format, true))
+  const isRelative = format === "relative" || format === "shortrelative"
+
+  onMount(() => {
```
Comment (robertvanhoesel) [inline src/components/widgets/FormatDate.tsx:84]: I replaced the `isServer` boolean with `onMount`, a hook that will only be called once, and only client side. I'm not sure if `isServer` is fully supported in Astro, but either way you want to only call these functions in a browser context.

I left your code structure as is but can be simplified after testing a bit more.

File: src/components/widgets/FormatDate.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  else return "now"
+}
+
+export interface FormatDateProps {
+  date: Date | string
+  format: DateFormat
+}
+export default function FormatDate({ date, format }: FormatDateProps) {
+  const d = parseUTC(date)
+  const [output, setOutput] = createSignal<string>(formatDate(d, format, true))
```
Comment (robertvanhoesel) [inline src/components/widgets/FormatDate.tsx:81]: We can initialize here at once

File: src/components/widgets/FormatDate.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  onMount(() => {
+    const update = () => {
+      const x = formatDate(d, format, false)
+      console.log("New date rendered:", x)
+      setOutput(x)
+    }
+
+    if (isRelative) {
+      // Dynamically updating display for client-side relative times
+      const interval = setInterval(update, 1000)
```
Comment (robertvanhoesel) [inline src/components/widgets/FormatDate.tsx:93]: My suggestion would be to only kick of this interval if the time difference is less than an hour. After that it is unlikely these calls will result in a textual change.

File: src/components/HypeCard.astro [hunk 4]
```diff
@@ -35,7 +35,7 @@ const website = (await getCollection("websites", ({ data }) => url?.includes(dat
       {
         date && (
           <p>
-            <Timeago date={date} />
+            <FormatDate date={date} format="relative" />
```
Comment (robertvanhoesel) [inline src/components/HypeCard.astro:38]:
```suggestion
            <FormatDate date={date} format="relative" client:load />
```

File: src/components/widgets/FormatDate.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      onCleanup(() => clearInterval(interval))
+    }
+
+    update()
+  })
+
+  return <>{output()}</>
```
Comment (robertvanhoesel) [inline src/components/widgets/FormatDate.tsx:100]: Perhaps would be nice for relative dates to print the normal formatted date into a tooltip

**Review by robertvanhoesel — CHANGES_REQUESTED** [review]

File: src/components/widgets/FormatDate.tsx (hunk 3 above)
Comment (acarapetis) [reply src/components/widgets/FormatDate.tsx:93]: Good call, will do that.

File: src/components/widgets/FormatDate.tsx (hunk 5 above)
Comment (acarapetis) [reply src/components/widgets/FormatDate.tsx:100]: Oops, Timeago had that and I forgot to port it over.

File: src/components/widgets/FormatDate.tsx (hunk 1 above)
Comment (acarapetis) [reply src/components/widgets/FormatDate.tsx:84]: Looks like that's fixed the bug, thanks!

**Comment by acarapetis** [comment]
@robertvanhoesel Thanks for the review! I've pushed some changes following your advice and removed the debugging code.

**Review by robertvanhoesel — APPROVED** [review]
Thanks, great work!

**Pull request merged**